File: backend/routers/transactions.py
from typing import Optional, List
from datetime import date, datetime
from uuid import UUID
from fastapi import APIRouter, Depends, HTTPException, status, Query, UploadFile, File
from sqlalchemy.orm import Session
from sqlalchemy import desc
from pydantic import BaseModel
import math
import csv
import io

from core.deps import get_db, get_current_user
from models.user import User
from models.transaction import Transaction
from schemas.transaction import (
    TransactionCreate,
    TransactionUpdate,
    TransactionResponse,
    TransactionListResponse,
    EXPENSE_CATEGORIES,
    INCOME_CATEGORIES
)

router = APIRouter(prefix="/transactions", tags=["transactions"])


class ImportResult(BaseModel):
    success: bool
    total_rows: int
    imported: int
    failed: int
    errors: List[dict]


def validate_category(type: str, category: str) -> bool:
    if type == "expense":
        return category.lower() in EXPENSE_CATEGORIES
    elif type == "income":
        return category.lower() in INCOME_CATEGORIES
    return False


@router.get("", response_model=TransactionListResponse)
def get_transactions(
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    type: Optional[str] = None,
    category: Optional[str] = None,
    start_date: Optional[date] = None,
    end_date: Optional[date] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    query = db.query(Transaction).filter(Transaction.user_id == current_user.id)

    if type:
        query = query.filter(Transaction.type == type)
    if category:
        query = query.filter(Transaction.category == category.lower())
    if start_date:
        query = query.filter(Transaction.date >= start_date)
    if end_date:
        query = query.filter(Transaction.date <= end_date)

    total = query.count()
    total_pages = math.ceil(total / limit) if total > 0 else 1

    transactions = (
        query
        .order_by(desc(Transaction.date), desc(Transaction.created_at))
        .offset((page - 1) * limit)
        .limit(limit)
        .all()
    )

    return TransactionListResponse(
        data=[TransactionResponse.model_validate(t) for t in transactions],
        total=total,
        page=page,
        limit=limit,
        total_pages=total_pages
    )


@router.post("", response_model=TransactionResponse, status_code=status.HTTP_201_CREATED)
def create_transaction(
    transaction_data: TransactionCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    if not validate_category(transaction_data.type, transaction_data.category):
        valid_categories = EXPENSE_CATEGORIES if transaction_data.type == "expense" else INCOME_CATEGORIES
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Kategori tidak valid untuk {transaction_data.type}. Pilihan: {', '.join(valid_categories)}"
        )

    transaction = Transaction(
        user_id=current_user.id,
        date=transaction_data.date,
        amount=transaction_data.amount,
        type=transaction_data.type,
        category=transaction_data.category.lower(),
        description=transaction_data.description
    )
    db.add(transaction)
    db.commit()
    db.refresh(transaction)

    return TransactionResponse.model_validate(transaction)


@router.put("/{transaction_id}", response_model=TransactionResponse)
def update_transaction(
    transaction_id: UUID,
    transaction_data: TransactionUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    transaction = (
        db.query(Transaction)
        .filter(Transaction.id == transaction_id, Transaction.user_id == current_user.id)
        .first()
    )

    if not transaction:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Transaksi tidak ditemukan"
        )

    update_data = transaction_data.model_dump(exclude_unset=True)

    # Validate category if type or category is being updated
    new_type = update_data.get("type", transaction.type)
    new_category = update_data.get("category", transaction.category)
    if "type" in update_data or "category" in update_data:
        if not validate_category(new_type, new_category):
            valid_categories = EXPENSE_CATEGORIES if new_type == "expense" else INCOME_CATEGORIES
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Kategori tidak valid untuk {new_type}. Pilihan: {', '.join(valid_categories)}"
            )

    for key, value in update_data.items():
        if key == "category" and value:
            value = value.lower()
        setattr(transaction, key, value)

    db.commit()
    db.refresh(transaction)

    return TransactionResponse.model_validate(transaction)


@router.delete("/{transaction_id}")
def delete_transaction(
    transaction_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    transaction = (
        db.query(Transaction)
        .filter(Transaction.id == transaction_id, Transaction.user_id == current_user.id)
        .first()
    )

    if not transaction:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Transaksi tidak ditemukan"
        )

    db.delete(transaction)
    db.commit()

    return {"message": "deleted"}


@router.post("/import", response_model=ImportResult)
async def import_transactions(
    file: UploadFile = File(...),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Import transactions from CSV file.

    Expected CSV format:
    date,amount,type,category,description
    2026-01-15,50000,expense,makan,Makan siang
    2026-01-16,2000000,income,gaji,Gaji bulanan

    - date: YYYY-MM-DD format
    - amount: positive integer
    - type: 'income' or 'expense'
    - category: valid category for the type
    - description: optional
    """
    if not file.filename.endswith('.csv'):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="File harus berformat CSV"
        )

    try:
        content = await file.read()
        # Try to decode with utf-8-sig (handles BOM), fallback to utf-8, then latin-1
        try:
            text = content.decode('utf-8-sig')  # Handles BOM
        except UnicodeDecodeError:
            try:
                text = content.decode('utf-8')
            except UnicodeDecodeError:
                text = content.decode('latin-1')

        # Parse CSV
        reader = csv.DictReader(io.StringIO(text))

        imported = 0
        failed = 0
        errors = []
        rows = list(reader)

        # Normalize column names (lowercase, strip whitespace)
        if rows:
            normalized_rows = []
            for row in rows:
                normalized_row = {k.lower().strip(): v for k, v in row.items() if k}
                normalized_rows.append(normalized_row)
            rows = normalized_rows

        for i, row in enumerate(rows, start=2):  # Start at 2 because row 1 is header
            try:
                # Parse and validate date
                date_str = row.get('date', row.get('tanggal', '')).strip()
                if not date_str:
                    raise ValueError("Tanggal kosong")

                # Support multiple date formats
                parsed_date = None
                for fmt in ['%Y-%m-%d', '%d/%m/%Y', '%d-%m-%Y', '%Y/%m/%d']:
                    try:
                        parsed_date = datetime.strptime(date_str, fmt).date()
                        break
                    except ValueError:
                        continue

                if not parsed_date:
                    raise ValueError(f"Format tanggal tidak valid: {date_str}")

                # Parse amount
                amount_str = row.get('amount', '').strip().replace('.', '').replace(',', '')
                if not amount_str:
                    raise ValueError("Amount kosong")
                amount = int(amount_str)
                if amount <= 0:
                    raise ValueError("Amount harus positif")

                # Parse type
                tx_type = row.get('type', '').strip().lower()
                if tx_type not in ['income', 'expense']:
                    raise ValueError(f"Type harus 'income' atau 'expense', got: {tx_type}")

                # Parse category
                category = row.get('category', '').strip().lower()
                if not category:
                    raise ValueError("Category kosong")

                if not validate_category(tx_type, category):
                    valid_cats = EXPENSE_CATEGORIES if tx_type == "expense" else INCOME_CATEGORIES
                    raise ValueError(f"Category '{category}' tidak valid untuk {tx_type}. Pilihan: {', '.join(valid_cats)}")

                # Parse description (optional)
                description = row.get('description', '').strip() or None

                # Create transaction
                transaction = Transaction(
                    user_id=current_user.id,
                    date=parsed_date,
                    amount=amount,
                    type=tx_type,
                    category=category,
                    description=description
                )
                db.add(transaction)
                imported += 1

            except Exception as e:
                failed += 1
                errors.append({
                    "row": i,
                    "data": dict(row),
                    "error": str(e)
                })

        # Commit all successful transactions
        if imported > 0:
            db.commit()

        return ImportResult(
            success=failed == 0,
            total_rows=len(rows),
            imported=imported,
            failed=failed,
            errors=errors[:10]  # Limit to first 10 errors
        )

    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Gagal memproses file: {str(e)}"
        )

File: backend/src/routes/transactions.js
const express = require('express');
const router = express.Router();
const multer = require('multer');
const { parse } = require('csv-parse/sync');
const prisma = require('../config/database');
const authMiddleware = require('../middleware/auth');
const { validateCategory, EXPENSE_CATEGORIES, INCOME_CATEGORIES } = require('../utils/categories');

const upload = multer({ storage: multer.memoryStorage() });

// Helper to convert BigInt to Number for JSON
const serializeTransaction = (t) => ({
  id: t.id,
  user_id: t.userId,
  date: t.date.toISOString().split('T')[0],
  amount: Number(t.amount),
  type: t.type,
  category: t.category,
  description: t.description,
  created_at: t.createdAt,
  updated_at: t.updatedAt,
});

// GET /transactions
router.get('/', authMiddleware, async (req, res, next) => {
  try {
    const { page = 1, limit = 20, type, category, start_date, end_date } = req.query;
    const pageNum = Math.max(1, parseInt(page));
    const limitNum = Math.min(100, Math.max(1, parseInt(limit)));
    const skip = (pageNum - 1) * limitNum;

    const where = { userId: req.user.id };

    if (type) where.type = type;
    if (category) where.category = category.toLowerCase();
    if (start_date || end_date) {
      where.date = {};
      if (start_date) where.date.gte = new Date(start_date);
      if (end_date) where.date.lte = new Date(end_date);
    }

    const [transactions, total] = await Promise.all([
      prisma.transaction.findMany({
        where,
        orderBy: [{ date: 'desc' }, { createdAt: 'desc' }],
        skip,
        take: limitNum,
      }),
      prisma.transaction.count({ where }),
    ]);

    res.json({
      data: transactions.map(serializeTransaction),
      total,
      page: pageNum,
      limit: limitNum,
      total_pages: Math.ceil(total / limitNum) || 1,
    });
  } catch (error) {
    next(error);
  }
});

// POST /transactions
router.post('/', authMiddleware, async (req, res, next) => {
  try {
    const { date, amount, type, category, description } = req.body;

    if (!date || !amount || !type || !category) {
      return res.status(400).json({ detail: 'Date, amount, type, dan category wajib diisi' });
    }

    if (amount <= 0) {
      return res.status(400).json({ detail: 'Amount harus lebih dari 0' });
    }

    if (!validateCategory(type, category)) {
      const validCategories = type === 'expense' ? EXPENSE_CATEGORIES : INCOME_CATEGORIES;
      return res.status(400).json({
        detail: `Kategori tidak valid untuk ${type}. Pilihan: ${validCategories.join(', ')}`
      });
    }

    const transaction = await prisma.transaction.create({
      data: {
        userId: req.user.id,
        date: new Date(date),
        amount: BigInt(amount),
        type,
        category: category.toLowerCase(),
        description: description || null,
      },
    });

    res.status(201).json(serializeTransaction(transaction));
  } catch (error) {
    next(error);
  }
});

// PUT /transactions/:id
router.put('/:id', authMiddleware, async (req, res, next) => {
  try {
    const { id } = req.params;
    const { date, amount, type, category, description } = req.body;

    const existing = await prisma.transaction.findFirst({
      where: { id, userId: req.user.id },
    });

    if (!existing) {
      return res.status(404).json({ detail: 'Transaksi tidak ditemukan' });
    }

    const newType = type || existing.type;
    const newCategory = category || existing.category;

    if ((type || category) && !validateCategory(newType, newCategory)) {
      const validCategories = newType === 'expense' ? EXPENSE_CATEGORIES : INCOME_CATEGORIES;
      return res.status(400).json({
        detail: `Kategori tidak valid untuk ${newType}. Pilihan: ${validCategories.join(', ')}`
      });
    }

    if (amount !== undefined && amount <= 0) {
      return res.status(400).json({ detail: 'Amount harus lebih dari 0' });
    }

    const updateData = {};
    if (date) updateData.date = new Date(date);
    if (amount) updateData.amount = BigInt(amount);
    if (type) updateData.type = type;
    if (category) updateData.category = category.toLowerCase();
    if (description !== undefined) updateData.description = description || null;

    const transaction = await prisma.transaction.update({
      where: { id },
      data: updateData,
    });

    res.json(serializeTransaction(transaction));
  } catch (error) {
    next(error);
  }
});

// DELETE /transactions/:id
router.delete('/:id', authMiddleware, async (req, res, next) => {
  try {
    const { id } = req.params;

    const existing = await prisma.transaction.findFirst({
      where: { id, userId: req.user.id },
    });

    if (!existing) {
      return res.status(404).json({ detail: 'Transaksi tidak ditemukan' });
    }

    await prisma.transaction.delete({ where: { id } });

    res.json({ message: 'deleted' });
  } catch (error) {
    next(error);
  }
});

// POST /transactions/import
router.post('/import', authMiddleware, upload.single('file'), async (req, res, next) => {
  try {
    if (!req.file) {
      return res.status(400).json({ detail: 'File tidak ditemukan' });
    }

    if (!req.file.originalname.endsWith('.csv')) {
      return res.status(400).json({ detail: 'File harus berformat CSV' });
    }

    const content = req.file.buffer.toString('utf-8');
    let rows;

    try {
      rows = parse(content, {
        columns: true,
        skip_empty_lines: true,
        trim: true,
      });
    } catch (parseError) {
      return res.status(400).json({ detail: `Gagal parse CSV: ${parseError.message}` });
    }

    let imported = 0;
    let failed = 0;
    const errors = [];

    for (let i = 0; i < rows.length; i++) {
      const row = rows[i];
      const rowNum = i + 2; // +2 because row 1 is header

      try {
        // Normalize keys
        const normalizedRow = {};
        for (const key of Object.keys(row)) {
          normalizedRow[key.toLowerCase().trim()] = row[key];
        }

        // Parse date
        const dateStr = (normalizedRow.date || normalizedRow.tanggal || '').trim();
        if (!dateStr) throw new Error('Tanggal kosong');

        let parsedDate;
        const dateFormats = [
          /^(\d{4})-(\d{2})-(\d{2})$/, // YYYY-MM-DD
          /^(\d{2})\/(\d{2})\/(\d{4})$/, // DD/MM/YYYY
          /^(\d{2})-(\d{2})-(\d{4})$/, // DD-MM-YYYY
        ];

        for (const fmt of dateFormats) {
          const match = dateStr.match(fmt);
          if (match) {
            if (fmt === dateFormats[0]) {
              parsedDate = new Date(match[1], match[2] - 1, match[3]);
            } else {
              parsedDate = new Date(match[3], match[2] - 1, match[1]);
            }
            break;
          }
        }

        if (!parsedDate || isNaN(parsedDate.getTime())) {
          throw new Error(`Format tanggal tidak valid: ${dateStr}`);
        }

        // Parse amount
        const amountStr = (normalizedRow.amount || '').replace(/[.,]/g, '').trim();
        if (!amountStr) throw new Error('Amount kosong');
        const amount = parseInt(amountStr);
        if (isNaN(amount) || amount <= 0) throw new Error('Amount harus positif');

        // Parse type
        const txType = (normalizedRow.type || '').toLowerCase().trim();
        if (!['income', 'expense'].includes(txType)) {
          throw new Error(`Type harus 'income' atau 'expense', got: ${txType}`);
        }

        // Parse category
        const category = (normalizedRow.category || '').toLowerCase().trim();
        if (!category) throw new Error('Category kosong');

        if (!validateCategory(txType, category)) {
          const validCats = txType === 'expense' ? EXPENSE_CATEGORIES : INCOME_CATEGORIES;
          throw new Error(`Category '${category}' tidak valid. Pilihan: ${validCats.join(', ')}`);
        }

        // Parse description
        const description = (normalizedRow.description || '').trim() || null;

        // Create transaction
        await prisma.transaction.create({
          data: {
            userId: req.user.id,
            date: parsedDate,
            amount: BigInt(amount),
            type: txType,
            category,
            description,
          },
        });

        imported++;
      } catch (err) {
        failed++;
        if (errors.length < 10) {
          errors.push({
            row: rowNum,
            data: row,
            error: err.message,
          });
        }
      }
    }

    res.json({
      success: failed === 0,
      total_rows: rows.length,
      imported,
      failed,
      errors,
    });
  } catch (error) {
    next(error);
  }
});

module.exports = router;
